core/market/decentralized/src/matcher/matcher.rs

```diff
@@ -1,45 +1,68 @@
-use std::path::Path;
-use std::sync::Arc;
 use thiserror::Error;
 use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
 
-use ya_client::model::market::{Demand, Offer, Proposal};
+use ya_client::model::market::Proposal;
 use ya_client::model::ErrorMessage;
 use ya_persistence::executor::DbExecutor;
 use ya_persistence::executor::Error as DbError;
+use ya_service_api_web::middleware::Identity;
 
 use crate::db::dao::*;
 use crate::db::models::Demand as ModelDemand;
 use crate::db::models::Offer as ModelOffer;
-use crate::db::models::SubscriptionId;
-use crate::db::*;
-use crate::migrations;
-use crate::protocol::{
-    Discovery, DiscoveryBuilder, DiscoveryError, DiscoveryInitError, PropagateOffer,
-    StopPropagateReason,
-};
-use crate::protocol::{OfferReceived, RetrieveOffers};
+use crate::db::models::{SubscriptionId, SubscriptionValidationError};
+use crate::protocol::{Discovery, DiscoveryInitError, Propagate, StopPropagateReason};
+use crate::protocol::{OfferReceived, OfferUnsubscribed, RetrieveOffers};
+use actix_web::{HttpResponse, ResponseError};
 
 #[derive(Error, Debug)]
 pub enum DemandError {
     #[error("Failed to save Demand. Error: {0}.")]
     SaveDemandFailure(#[from] DbError),
     #[error("Failed to remove Demand [{1}]. Error: {0}.")]
-    RemoveDemandFailure(DbError, String),
+    RemoveDemandFailure(DbError, SubscriptionId),
     #[error("Demand [{0}] doesn't exist.")]
-    DemandNotExists(String),
+    DemandNotExists(SubscriptionId),
+}
+
+impl ResponseError for DemandError {
+    fn error_response(&self) -> HttpResponse {
+        match self {
+            DemandError::DemandNotExists(e) => {
+                HttpResponse::NotFound().json(ErrorMessage::new(self.to_string()))
+            }
+            _ => HttpResponse::InternalServerError().json(ErrorMessage::new(self.to_string())),
+        }
+    }
 }
```

Comment:
Http stuff shouldn't leak into our internals in my opinion.
I would move it to api.rs.

Reply:
will do